pkg/tsdb/azuremonitor/applicationinsights-datasource.go

```diff
@@ -0,0 +1,592 @@
+	var data ApplicationInsightsQueryResponse
+	err := json.Unmarshal(body, &data)
+	if err != nil {
+		azlog.Error("Failed to unmarshal AzureMonitor response", "error", err, "body", string(body))
+		return nil, nil, err
+	}
+
+	type Metadata struct {
+		Columns []string `json:"columns"`
+	}
+
+	meta := Metadata{}
+
+	for _, t := range data.Tables {
+		if t.Name == "PrimaryResult" {
+			timeIndex, valueIndex, segmentIndex := -1, -1, -1
+			meta.Columns = make([]string, 0)
+			for i, v := range t.Columns {
+				meta.Columns = append(meta.Columns, v.Name)
+				switch v.Name {
+				case query.TimeColumnName:
+					timeIndex = i
+				case query.ValueColumnName:
+					valueIndex = i
+				case query.SegmentColumnName:
+					segmentIndex = i
+				}
+			}
+
+			if timeIndex == -1 {
+				azlog.Info("no time column specified, returning existing columns, no data")
+				return nil, simplejson.NewFromAny(meta), nil
+			}
+
+			if valueIndex == -1 {
+				azlog.Info("no value column specified, returning existing columns, no data")
+				return nil, simplejson.NewFromAny(meta), nil
+			}
+
+			var getPoints func([]interface{}) *tsdb.TimeSeriesPoints
+			slice := tsdb.TimeSeriesSlice{}
+			if segmentIndex == -1 {
+				legend := formatApplicationInsightsLegendKey(query.Alias, query.ValueColumnName, "", "")
+				series := tsdb.NewTimeSeries(legend, []tsdb.TimePoint{})
+				slice = append(slice, series)
+				getPoints = func(row []interface{}) *tsdb.TimeSeriesPoints {
+					return &series.Points
+				}
+			} else {
+				mapping := map[string]*tsdb.TimeSeriesPoints{}
+				getPoints = func(row []interface{}) *tsdb.TimeSeriesPoints {
+					segment := fmt.Sprintf("%v", row[segmentIndex])
+					if points, ok := mapping[segment]; ok {
+						return points
+					}
+					legend := formatApplicationInsightsLegendKey(query.Alias, query.ValueColumnName, query.SegmentColumnName, segment)
+					series := tsdb.NewTimeSeries(legend, []tsdb.TimePoint{})
+					slice = append(slice, series)
+					mapping[segment] = &series.Points
+					return &series.Points
+				}
+			}
+
+			for _, r := range t.Rows {
+				timeStr, ok := r[timeIndex].(string)
+				if !ok {
+					return nil, simplejson.NewFromAny(meta), errors.New("invalid time value")
+				}
+				timeValue, err := time.Parse(time.RFC3339Nano, timeStr)
+				if err != nil {
+					return nil, simplejson.NewFromAny(meta), err
+				}
+
+				var value float64
+				if value, err = getFloat(r[valueIndex]); err != nil {
+					return nil, simplejson.NewFromAny(meta), err
+				}
+
+				points := getPoints(r)
+				*points = append(*points, tsdb.NewTimePoint(null.FloatFrom(value), float64(timeValue.Unix()*1000)))
+			}
+
+			return slice, simplejson.NewFromAny(meta), nil
+		}
+	}
+
+	return nil, nil, errors.New("could not find table")
+}
+
+func (e *ApplicationInsightsDatasource) parseTimeSeriesFromMetrics(body []byte, query *ApplicationInsightsQuery) (tsdb.TimeSeriesSlice, error) {
+	doc, err := simplejson.NewJson(body)
```

Comment:
It would be better if the response from Application Insights metrics could be typed instead. This could be difficult as the format of the response changes a lot. We could also merge first and fix afterwards.

Reply:
I initially tried that, but there are two things that I couldn't resolve.  Firstly, the answer varies _wildly_ depending in if you are asking for A) intervaled values or just a single datapoint and B) if you've added a dimension or not. Secondly, the _keys_ of the JSON values are intermixed. The result looks like
```json
{
  "start": "2019...",
  "end": "2019...",
  "NAME-OF-METRIC": ...
```

And I couldn't figure out a good typed way to handle "extra keys" that didn't just drop them.

I'm new to Go, so it's possible I missed something, and I'd be happy to change it if those problems are surmountable.

Reply:
Yes, I guessed that was the reason you went with SimpleJson. I can have a look and see if it is possible to do it but I can do that after merging. We did manage to do it for other datasources but none of them was as fickle as the App Insights data format.